### PR TITLE
Making Win32Exception.ToString() method contain NativeErrorCode

### LINKED ISSUE #24892: Result string of Win32Exception.ToString() method missing Win32Exception.NativeErrorCode
When we working with Win API via P/Invoke, it is common practice to check the result of called function and throw new [Win32Exception()](https://github.com/dotnet/corefx/blob/9eafffc6054c8f039d788465621bcfd49a379ca1/src/Microsoft.Win32.Primitives/src/System/ComponentModel/Win32Exception.cs#L21). We can pass error code explicitly, or the default constructor will catch the last one using [Marshal.GetLastWin32Error()](https://github.com/dotnet/coreclr/blob/da8b7181dbe362fbfe64c7ef72ff3a4a7db68f24/src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs#L839). When we log the exception, the result string does not contain root *Win32Exception.NativeErrorCode*:

```csharp
var winError = new System.ComponentModel.Win32Exception(0x35);
/*
winError.ToString() provides:
"System.ComponentModel.Win32Exception (0x80004005): The network path was not found"
*/
```

The *0x80004005* referes to the common HResult.E_FAIL, which is _Unspecified error_ in winerror.h; That's wrong (IMHO)! We have provided exact error code, it is well specified.

Expected result:

```csharp
/*
winError.ToString() provides:
"System.ComponentModel.Win32Exception (0x80070035): The network path was not found"
OR
"System.ComponentModel.Win32Exception (0x00000035): The network path was not found"
*/
```

The last one logs the number we passed, the first one is the result of [Marshal.GetHRForLastWin32Error()](https://github.com/dotnet/coreclr/blob/da8b7181dbe362fbfe64c7ef72ff3a4a7db68f24/src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs#L852)

### PR DESCRIPTION
**Pull request by stukselbax**
Override Win32Exception.ToString() method so that resulting string contain the Win32Exception.NativeErrorCode value. Adding this checks to the unit test. 

Fixes #24892

### CODE REVIEW

**Comment by benaadams** [comment]
Would it be better to leave `.ToString()` as is and just change the two calls to the base class?

```csharp
public Win32Exception(int error, string message)
 : base($"({error:R}): {message}")

public Win32Exception(string message, Exception innerException)
 : base($"({Marshal.GetLastWin32Error():R}): {message}", innerException)
```

**Comment by benaadams** [comment]
Don't see how it would produce any different output to this change? Adding in the method bodies (which would be unchanged)
```csharp
public Win32Exception(int error, string message)
 : base($"{error:R}: {message}")
{
    NativeErrorCode = error;
}

public Win32Exception(string message, Exception innerException)
 : base($"{Marshal.GetLastWin32Error():R}: {message}", innerException)
{
    NativeErrorCode = Marshal.GetLastWin32Error();
}
```
Its just passing the `NativeErrorCode` into the base message? I might be missing something?

**Comment by stukselbax** [comment]
Look at [ExternalException.ToString()](https://github.com/dotnet/coreclr/blob/da66c3a146563d5841502a8dec76fe082ee9fef6/src/mscorlib/shared/System/Runtime/InteropServices/ExternalException.cs#L64) method base class and you will notice that it first produce the name of class, and then - HResult, and later - the provided message:

```csharp
        public override string ToString()
        {
            string message = Message;
            string className = GetType().ToString();

            string s = className + " (0x" + HResult.ToString("X8", CultureInfo.InvariantCulture) + ")";
            //...
            return s;
        }
```

**Comment by benaadams** [comment]
Ah... missed what the parent class was

File: src/Microsoft.Win32.Primitives/src/System/ComponentModel/Win32Exception.cs [hunk 1]
```diff
@@ -70,5 +72,39 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
...
+
+            Exception innerException = InnerException;
+
+            if (innerException != null)
+            {
+                s.Append(" ---> ");
+                s.Append(innerException.ToString());
+            }
+
+            if (StackTrace != null)
```
Comment (stephentoub) [inline src/Microsoft.Win32.Primitives/src/System/ComponentModel/Win32Exception.cs]: We should cache this in a local, e.g.
```C#
string stackTrace = StackTrace;
if (stackTrace != null)
{
    s.AppendLine().Append(stackTrace);
}
```

**Comment by stukselbax** [comment]
I'm also leave the **NativeErrorCode** in decimal, because I hope that it is expected value for the vast majority of developers.

**Comment by danmosemsft** [comment]
Could you please look at the CI failures? Not that the same tests are run against .NET Framework (this is the netfx run) so the tests may need conditionalizing using a ConditionalFact attribute

**Comment by stukselbax** [comment]
@danmosemsft I have looked at the errors, and for the first time I couldn't imagine that I could broke some tests, specially with the following failure in the test **System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests**:

> Error during equality check of type System.Net.WebSockets.WebSocketException
> Expected: True
> Actual:   False

Looking closer I have found out that WebSocketException is inhireted from Win32Exception. Wow! That is unexpectable, imho. I'll check what I can do to fix it.

**Comment by stukselbax** [comment]
The test was not passed due to unequality of string representations of the **WebSocketException**'s - source exception has **NativeErrorCode** equals to 0, serialized exception has value 2 (after deserialization).  I think this was introduced in #24427, because the deserialization constructor of **Win32Exception** was changed:

```csharp
    protected Win32Exception(SerializationInfo info, StreamingContext context) : base(info, context)
    {
        NativeErrorCode = info.GetInt32(nameof(NativeErrorCode));
    }
```

And also those test data was added. The problem data is the second blob for **WebSocketException**.
@ViktorHofer How can I produce correct value for test data here? Is it enough?

```csharp
var webSocketException = new WebSocketException(WebSocketError.Success, 100, "message", exception);
webSocketException = PopulateException(webSocketException); // As stored in the SampleData
var binaryFormatter = new BinaryFormatter();
binaryFormatter.AssemblyFormat = assemblyStyle; // Full?
byte[] rawData;
using (var serializedStream = new MemoryStream(raw))
{
    binaryFormatter.Serialize(serializedStream, webSocketException );
    rawData = serializedStream.ToArray()
}

string result = Convert.ToBase64String(rawData);
```

For some reason **WebSocketException** uses **HResult** instead of **NativeErrorCode**. So, currently I'm doubt that my changes are acceptable. Should I change **ToString()** method to check corner cases for **NativeErrorCode**, and if it is 0, return **base.ToString()**;

**Comment by ViktorHofer** [comment]
What I normall do:
1. Change the test code (in your case WebSocketException) here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs
2. Uncomment this line: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs#L36
3. Run the UpdateBlobs test for both netcoreapp and netfx: `msbuild /t:RebuildAndTest /p:xunitoptions="-method System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.UpdateBlobs"`
`msbuild /t:RebuildAndTest /p:TargetGroup=netfx /p:xunitoptions="-method System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.UpdateBlobs"`
4. Comment the line again: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs#L36

If you c

**Comment by stukselbax** [comment]
The problem I'm faced with relates to **WebSocketException**. Actually, it is inherited from **Win32Exception**, but the API it uses (on Windows) operates with HResult error code format, and when initialized, it has code path where **NativeErrorCode** is set, and other path where **NativeErrorCode** are not set, and some times both **NativeErrorCode** and **HResult** can be set. So, in test that was failed, with new **.ToString()** method was introduced, the result will be:

```csharp
var webSocketException = new WebSocketException(WebSocketError.Success, 100, "message", exception);
webSocketException.ToString();

// output as follows:
// System.Net.WebSockets.WebSocketException (0): message ...
```

Which does not looks great. Other inheriters of **Win32Exception** are more tolerant and use **NativeErrorCode** as expected.

That lead us to add some analyze of **NativeErrorCode**, **HResult**, and decide what should be included in the string. Should we print both of them, if they are not match? Can we omit the generic E_FAIL (0x80004005)?

**Comment by danmosemsft** [comment]
@stukselbax what about overriding ToString in WebSocketException to preserve the current format for it?

**Comment by stukselbax** [comment]
@danmosemsft that will make me sad because of copy-paste solution involved. I have tried to change the way of generating result string. The worst case is when
```csharp
 (HResult != E_FAIL && NativeErrorCode != 0 && HResult != NativeErrorCode)
``` 
This will rarely happen (I hope), but it is better to print both HResult & NativeErrorCode, cause they are not default.

**Comment by stukselbax** [comment]
Currently the test I've added is failed for **NETFX x86 Release Build**.  AFAIK, **NETFX** - is [full .net framework](https://github.com/dotnet/corefx/wiki/Build-and-run-tests), and as @danmosemsft already mentioned - it should be _conditionalizing_, but I don't want to exclude the whole test, only the part which fail.

**Comment by danmosemsft** [comment]
@stukselbax what does ToString look like now for Win32Exception and for WebSocketException vs. what it did before?

Looks like we are just about done.

**Comment by stukselbax** [comment]
Compare list:

```csharp
LastError 53
    new Win32Exception()
    // Before: System.ComponentModel.Win32Exception (0x80004005): The network path was not found
    // After:  System.ComponentModel.Win32Exception (53): The network path was not found

    new Win32Exception(2)
    // Before: System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
    // After:  System.ComponentModel.Win32Exception (2): The system cannot find the file specified

    new Win32Exception(message)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message
    // After:  System.ComponentModel.Win32Exception (0x80004005): message

    new Win32Exception(2, message)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message
    // After:  System.ComponentModel.Win32Exception (2): message

    new Win32Exception(message, inner)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message ---> System.Exception: message
    // After:  System.ComponentModel.Win32Exception (0x80004005): message ---> System.Exception: message

    new SocketException()
    // Before: System.Net.Sockets.SocketException (0x80004005): The operation completed successfully
    // After:  System.Net.Sockets.SocketException (0x80004005): The operation completed successfully

    new SocketException(2);
    // Before: System.Net.Sockets.SocketException (0x80004005): The system cannot find the file specified
    // After:  System.Net.Sockets.SocketException (2): The system cannot find the file specified

    new WebSocketException()
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The operation completed successfully
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The operation completed successfully

    new WebSocketException(2)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The system cannot find the file specified
    // After:  System.Net.WebSockets.WebSocketException (2): The system cannot find the file specified

    new WebSocketException(message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message

    new WebSocketException(WebSocketError.InvalidMessageType)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.

    new WebSocketException(2, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message

    new WebSocketException(2, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (2): message

    new WebSocketException(message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel. ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel. ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, 2)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.

    new WebSocketException(WebSocketError.InvalidMessageType, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message

    new WebSocketException(WebSocketError.InvalidMessageType, 2, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message

    new WebSocketException(WebSocketError.InvalidMessageType, message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, 2, message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message

LastError 0x80070020
    new Win32Exception()
    // Before: System.ComponentModel.Win32Exception (0x80004005): The process cannot access the file because it is being used by another process
    // After:  System.ComponentModel.Win32Exception (0x80070020): The process cannot access the file because it is being used by another process

    new Win32Exception(0x80070021)
    // Before: System.ComponentModel.Win32Exception (0x80004005): The process cannot access the file because another process has locked a portion of the file
    // After:  System.ComponentModel.Win32Exception (0x80070021): The process cannot access the file because another process has locked a portion of the file

    new Win32Exception(message)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message
    // After:  System.ComponentModel.Win32Exception (0x80004005): message

    new Win32Exception(0x80070021, message)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message
    // After:  System.ComponentModel.Win32Exception (0x80070021): message

    new Win32Exception(message, inner)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message ---> System.Exception: message
    // After:  System.ComponentModel.Win32Exception (0x80004005): message ---> System.Exception: message

    new SocketException()
    // Before: System.Net.Sockets.SocketException (0x80004005): The operation completed successfully
    // After:  System.Net.Sockets.SocketException (0x80004005): The operation completed successfully

    new SocketException(0x80070021);
    // Before: System.Net.Sockets.SocketException (0x80004005): The process cannot access the file because another process has locked a portion of the file
    // After:  System.Net.Sockets.SocketException (0x80070021): The process cannot access the file because another process has locked a portion of the file

    new WebSocketException()
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The operation completed successfully
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The operation completed successfully

    new WebSocketException(0x80070021)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): The process cannot access the file because another process has locked a portion of the file
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): The process cannot access the file because another process has locked a portion of the file

    new WebSocketException(message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message

    new WebSocketException(WebSocketError.InvalidMessageType)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.

    new WebSocketException(0x80070021, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message

    new WebSocketException(0x80070021, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): message
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): message

    new WebSocketException(message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel. ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel. ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, 0x80070021)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): The received  message type is invalid after calling WebSocket.CloseAsync. WebSocket.CloseAsync should only be used if no more data is expected from the remote endpoint. Use 'WebSocket.CloseOutputAsync' instead to keep being able to receive data but close the output channel.

    new WebSocketException(WebSocketError.InvalidMessageType, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message

    new WebSocketException(WebSocketError.InvalidMessageType, 0x80070021, message)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): message
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): message

    new WebSocketException(WebSocketError.InvalidMessageType, message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): message ---> System.Exception: message

    new WebSocketException(WebSocketError.InvalidMessageType, 0x80070021, message, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80070021): message ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80070021): message ---> System.Exception: message
```

**Comment by stukselbax** [comment]
cc @danmosemsft @karelz
What else should I do?

**Comment by danmosemsft** [comment]
@stukselbax my bad, I missed this was pending.

It looks odd that the WebSocketException messages where there is a nonzero native error code (2, here) show E_FAIL instead of 2. I wonder whether they should get the same improvement as Win32Exception. However, you did not change this behavior.

For the LastError 0x80070020 case, I'm having trouble seeing why I see 0x80004005 here:
    new Win32Exception(message)
    // Before: System.ComponentModel.Win32Exception (0x80004005): message
    // After:  System.ComponentModel.Win32Exception (0x80004005): message
It seems from the code that HResult would remain 0x80004005 and so it would display 0x80070020 as I would expect.

Do we have enough tests for WebSocketException and SocketException to protect the messages you've pasted above?

**Comment by danmosemsft** [comment]
@stukselbax  thoughts? Thanks for your patience.

**Comment by stukselbax** [comment]
@danmosemsft sorry for late response. 

```csharp
    new WebSocketException(2, inner)
    // Before: System.Net.WebSockets.WebSocketException (0x80004005): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message
    // After:  System.Net.WebSockets.WebSocketException (0x80004005): An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Exception: message
```

this is because [WebSocketException(int, innerException)](https://github.com/dotnet/corefx/blob/e3f74deb999571d3f34a6b2b8011caebcf06dbfc/src/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketException.cs) constructor is implemented that way:

```csharp
        public WebSocketException(int nativeError, Exception innerException)
            : base(SR.net_WebSockets_Generic, innerException)
        {
            _webSocketErrorCode = !Succeeded(nativeError) ? WebSocketError.NativeError : WebSocketError.Success;
            SetErrorCodeOnError(nativeError);
        }
```

it does not pass the nativeError to the base class, and will not set the HResult = 2, because it is not an HResult.

For the second concern - it seems that method which generates the diff is incorrect, I'll check it once again.

I don't think that tests which checks the messages for exception you have specified exists r required. My initial point was about Win32Exception, and SocketException currently satisfy my expectation as well. For the WebSocketException - the Win API which is used for them returns HResult as error code, and probably the messages preserved in this commit are not relevant.

**Comment by stukselbax** [comment]
@danmosemsft it really was a bug in code which generates the output. I have edited comment comparing behavior before and after, please, take a look one more time.

**Comment by danmosemsft** [comment]
LGTM. Thanks @stukselbax . 

It seems wrong to me that in some cases WSE uses the native code and some cases it does not. It may be that this is tied to how it is used by the web socket code. Perhaps that should be cleaned up.

**Pull request merged**